oxidized-transformers/src/models/roberta/embeddings.rs

```diff
@@ -0,0 +1,112 @@
+use candle_core::{DType, Tensor};
+use candle_nn::VarBuilder;
+use snafu::{ResultExt, Snafu};
+
+use crate::architectures::{BuildEmbeddings, Embeddings};
+use crate::error::BoxedError;
+use crate::layers::transformer::{TransformerEmbeddingsConfig, TransformerEmbeddingsError};
+
+#[derive(Debug)]
+/// RoBERTa embeddings configuration.
+pub struct RobertaEmbeddingsConfig {
+    padding_id: u32,
+    transformer_embeddings: Box<dyn BuildEmbeddings>,
+}
+
+impl Default for RobertaEmbeddingsConfig {
+    fn default() -> Self {
+        Self {
+            padding_id: 1,
+            transformer_embeddings: Box::<TransformerEmbeddingsConfig>::default(),
+        }
+    }
+}
+
+impl RobertaEmbeddingsConfig {
+    /// Padding token id.
+    ///
+    /// Default: 1
+    pub fn padding_id(mut self, padding_id: u32) -> Self {
+        self.padding_id = padding_id;
+        self
+    }
+
+    /// Transformer embeddings configuration.
+    ///
+    /// Default: `TransformerEmbeddingsConfig::default()`
+    pub fn transformer_embeddings(
+        mut self,
+        transformer_embeddings: Box<dyn BuildEmbeddings>,
+    ) -> Self {
+        self.transformer_embeddings = transformer_embeddings;
+        self
+    }
+}
+
+impl BuildEmbeddings for RobertaEmbeddingsConfig {
+    fn build(&self, vb: VarBuilder) -> Result<Box<dyn Embeddings>, BoxedError> {
+        Ok(Box::new(RobertaEmbeddings {
+            inner: self.transformer_embeddings.build(vb)?,
+            padding_id: self.padding_id,
+        }))
+    }
+}
+
+/// RoBERTa embeddings errors.
+#[derive(Debug, Snafu)]
+pub enum RobertaEmbeddingsError {
+    #[snafu(display("Error computing transformer embeddings"))]
+    TransformerEmbeddings { source: TransformerEmbeddingsError },
+
+    #[snafu(display("Cannot compute token mask"))]
+    ComputeMask { source: candle_core::Error },
+
+    #[snafu(display("Cannot compute token positions"))]
+    ComputePositions { source: candle_core::Error },
+
+    #[snafu(display("Cannot convert padding id to tensor"))]
+    PaddingIdToTensor { source: candle_core::Error },
+}
+
+/// RoBERTa (Li et al., 2019) embeddings.
+pub struct RobertaEmbeddings {
+    inner: Box<dyn Embeddings>,
+    padding_id: u32,
+}
+
+impl Embeddings for RobertaEmbeddings {
+    fn forward(
+        &self,
+        piece_ids: &Tensor,
+        train: bool,
+        positions: Option<&Tensor>,
+        type_ids: Option<&Tensor>,
+    ) -> Result<Tensor, BoxedError> {
+        let positions = match positions {
+            Some(positions) => positions.clone(),
```

Comment:
Seeing how there's only one more statement after the position tensor calculation, wouldn't it make sense to elide this clone and directly call `self.inner` inside each match arm?

Reply:
Hmmm, then these `forward` calls are duplicated, which is also not really nice? `clone` is just an `Arc` refcount increment, so the cost is negligible compared to the work done in the `forward`. I initially had some of these cases wrapped in `Cow`, but then removed that, since it's not really worth sacrificing readability over one bump.

Reply:
Fair enough.